pkg/endpoint/bpf.go

```diff
@@ -822,7 +647,183 @@ func (e *Endpoint) realizeBPFState(regenContext *RegenerationContext) (compilati
+		e.PolicyMap, _, err = policymap.OpenMap(e.PolicyMapPathLocked())
+		if err != nil {
+			return err
+		}
+		// Clean up map contents
+		e.getLogger().Debug("flushing old PolicyMap")
+		err = e.PolicyMap.Flush()
+		if err != nil {
+			return err
+		}
+
+		// Also reset the in-memory state of the realized state as the
+		// BPF map content is guaranteed to be empty right now.
+		e.realizedMapState = make(PolicyMapState)
+	}
+
+	// Only generate & populate policy map if a security identity is set up for
+	// this endpoint.
+	if e.SecurityIdentity != nil {
+		stats.policyCalculation.Start()
+		err = e.regeneratePolicy(owner)
+		stats.policyCalculation.End(err == nil)
+		if err != nil {
+			return fmt.Errorf("unable to regenerate policy for '%s': %s", e.PolicyMap.String(), err)
+		}
+
+		_ = e.updateAndOverrideEndpointOptions(nil)
+
+		// Synchronously try to update PolicyMap for this endpoint. If any
+		// part of updating the PolicyMap fails, bail out and do not generate
+		// BPF. Unfortunately, this means that the map will be in an inconsistent
+		// state with the current program (if it exists) for this endpoint.
+		// GH-3897 would fix this by creating a new map to do an atomic swap
+		// with the old one.
+		stats.mapSync.Start()
+		err := e.syncPolicyMap()
+		stats.mapSync.End(err == nil)
+		if err != nil {
+			return fmt.Errorf("unable to regenerate policy because PolicyMap synchronization failed: %s", err)
+		}
+
+		// Configure the new network policy with the proxies.
+		stats.proxyPolicyCalculation.Start()
+		var networkPolicyRevertFunc revert.RevertFunc
+		err, networkPolicyRevertFunc = e.updateNetworkPolicy(owner, datapathRegenCtxt.proxyWaitGroup)
+		stats.proxyPolicyCalculation.End(err == nil)
+		if err != nil {
+			return err
+		}
+
+		datapathRegenCtxt.revertStack.Push(networkPolicyRevertFunc)
+	}
+
+	stats.proxyConfiguration.Start()
+	var finalizeFunc revert.FinalizeFunc
+	var revertFunc revert.RevertFunc
+	// Walk the L4Policy to add new redirects and update the desired policy map
+	// state to set the newly allocated proxy ports.
+	var desiredRedirects map[string]bool
+	if e.DesiredL4Policy != nil {
+		desiredRedirects, err, finalizeFunc, revertFunc = e.addNewRedirects(owner, e.DesiredL4Policy, datapathRegenCtxt.proxyWaitGroup)
+		if err != nil {
+			stats.proxyConfiguration.End(false)
+			return err
+		}
+		datapathRegenCtxt.finalizeList.Append(finalizeFunc)
+		datapathRegenCtxt.revertStack.Push(revertFunc)
+	}
+
+	// At this point, traffic is no longer redirected to the proxy for
+	// now-obsolete redirects, since we synced the updated policy map above.
+	// It's now safe to remove the redirects from the proxy's configuration.
+	finalizeFunc, revertFunc = e.removeOldRedirects(owner, desiredRedirects, datapathRegenCtxt.proxyWaitGroup)
+	datapathRegenCtxt.finalizeList.Append(finalizeFunc)
+	datapathRegenCtxt.revertStack.Push(revertFunc)
+	stats.proxyConfiguration.End(true)
+
+	stats.prepareBuild.Start()
+	defer func() {
```

Comment:
This is the only change in the code - I moved `prepareBuild.End` to be within this function instead of included with `waitingForCTClean`.